File: 17. Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        coll={'2':'abc','3':'def','4':'ghi','5':'jkl','6':'mno','7':'pqrs','8':'tuv','9':'wxyz'}
        if len(digits) == 0:
            return ''
        res=[]
        
        def rec(idx,ans):
            if len(ans)==len(digits):
                res.append(ans)
                return
            for q in coll[digits[idx]]:
                ans+=q
                rec(idx+1,ans)
                ans=ans[:-1]
            
        rec(0,"")
        return res
        

File: 17. Letter Combinations of a Phone Number.java
class Solution {
	static String[] arr = { ".;","", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
	
	public List<String> letterCombinations(String digits) {
		
		if (digits.length() == 0) {
			return new ArrayList<String>();
		}
		
		return helper(digits);
	}

	public List<String> helper(String digits) {
		if (digits.length() == 0) {
			List<String> temp = new ArrayList<>();
			temp.add("");
			return temp;
		}

		char ch = digits.charAt(0);
		String ros = digits.substring(1);
		List<String> res = helper(ros);
		List<String> ans = new ArrayList<>();

		String chcode = arr[ch - '0'];

		for (int i = 0; i < chcode.length(); i++) {
			char chc = chcode.charAt(i);
			for (String st : res)
			{
				ans.add(chc + st);
			}
			}
		return ans;
	}

}
